modules/internal/Bytes.chpl

```diff
@@ -0,0 +1,473 @@
+
+/*
+ * Copyright 2004-2019 Cray Inc.
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*
+The following documentation shows functions and methods used to
+manipulate and process Chapel bytes objects.
+
+Casts from bytes to a Numeric Type
+-----------------------------------
+
+This module supports casts from :record:`bytes` to numeric types. Such casts
+will convert the bytes to the numeric type and throw an error if the bytes
+is invalid. For example:
+
+.. code-block:: chapel
+
+  var number = "a":int;
+
+throws an error when it is executed, but
+
+.. code-block:: chapel
+
+  var number = "1":int;
+
+stores the value ``1`` in ``number``.
+
+To learn more about handling these errors, see the
+:ref:`Error Handling technical note <readme-errorHandling>`.
+ */
+module Bytes {
+  use ChapelStandard;
+
+  // to follow String.chpl, and maybe limit the index to smaller integers
+  type byteIndex = int; 
+
+  type byteType = uint(8);
+  type bufferType = c_ptr(byteType);
+
+  record bytes {  // _bytes when/if there is compiler support
+    pragma "no doc"
+    var len: int = 0;
+    pragma "no doc"
+    var buff: bufferType = nil;
+
+    pragma "no doc"
+    proc init() {
+
+    }
+
+    /*
+      Initialize a new :record:`bytes` from ``b``. If ``isowned`` is set to
+      ``true`` then ``b`` will be fully copied into the new instance. If it is
+      ``false`` a shallow copy will be made such that any in-place modifications
+      to the new bytes may appear in ``b``. It is the responsibility of the user
+      to ensure that the underlying buffer is not freed while being used as part
+      of a shallow copy.
+     */
+    proc init(b: bytes, isowned: bool = true) {
```

Comment:
I would expect the API to include ways to construct a `bytes` from a `c_string` and a length or from a `c_ptr` and a length.